grass/kernel.c

```diff
@@ -174,12 +181,32 @@ static void proc_syscall() {
 
     switch (type) {
     case SYS_RECV:
-        proc_recv(sc);
+        rc = proc_recv(sc);
         break;
     case SYS_SEND:
-        proc_send(sc);
+        rc = proc_send(sc);
+        break;
+    case SYS_WAIT:
+        waiting++;
+        if (wakeup) rc = waiting = wakeup = 0;
+        break;
+    case SYS_EXIT:
+        proc_free(curr_pid);
+        if (waiting) wakeup++;
         break;
     default:
         FATAL("proc_syscall: got unknown syscall type=%d", type);
     }
+
+    if (rc == -1)
+        sc->type = type; // Failure, Retry Request
+    else
+        proc_set_runnable(curr_pid); // Either Success or Error, Move to User Space
```

Comment:
Instead of using `curr_pid`, maybe we should make `pid` an argument of `proc_syscall` since this function may be handling a kernel function that was invoked by another process earlier in time.

Reply:
Since there were usages of more items in the process' PCB than just their `pid`, a pointer to the process' PCB entry became an argument to `proc_syscall` as `proc_syscall(struct process *proc)`. Let me know if you think this is a needed change, or only `pid` is necessary.